rmw_zenoh_cpp/src/rmw_publisher.cpp

```diff
@@ -118,44 +111,49 @@ rmw_create_publisher(
   }
 
   publisher->data = static_cast<rmw_publisher_data_t *>(
-    allocator->allocate(sizeof(rmw_publisher_data_t), allocator->state)
-  );
+      allocator->allocate(sizeof(rmw_publisher_data_t), allocator->state));
   if (!publisher->data) {
     RMW_SET_ERROR_MSG("failed to allocate publisher data");
     allocator->deallocate(publisher->data, allocator->state);
+
+    allocator->deallocate(const_cast<char *>(publisher->topic_name), allocator->state);
     allocator->deallocate(publisher, allocator->state);
     return nullptr;
   }
 
   publisher->options = *publisher_options;
 
   // CREATE PUBLISHER MEMBERS ==================================================
+  // Get typed pointer to implementation specific publisher data struct
+  auto publisher_data = static_cast<rmw_publisher_data_t *>(publisher->data);
+
   // Init type support callbacks
   auto callbacks = static_cast<const message_type_support_callbacks_t *>(type_support->data);
 
   // Create Zenoh resource
   ZNSession * s = node->context->impl->session;
-  size_t zn_topic_id = zn_declare_resource(s, publisher->topic_name);
 
-  // Get typed pointer to implementation specific publisher data struct
-  auto publisher_data = static_cast<rmw_publisher_data_t *>(publisher->data);
+  // NOTE(CH3): This topic ID only unique WITHIN this process!
+  //
+  // Another topic on another process might clash with the ID on this process, even within the
+  // same Zenoh network! It is not a UUID!!
+  publisher_data->zn_topic_id_ = zn_declare_resource(s, publisher->topic_name);
```

Comment:
```suggestion
  // The topic ID must be unique within a single process, but separate processes can reuse IDs, even in the same
  // Zenoh network, because the ID is never transmitted over the wire.
  // Conversely, the ID used in two communicating processes cannot be used to determine if they are using the same
  // topic or not.
```